// avalanche/typings/tests/apis/health/api.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jest_mock_axios_1 = __importDefault(require("jest-mock-axios"));
const src_1 = require("src");
const api_1 = require("../../../src/apis/health/api");
describe("Health", () => {
    const ip = "127.0.0.1";
    const port = 9650;
    const protocol = "https";
    const avalanche = new src_1.Avalanche(ip, port, protocol, 12345, undefined, undefined, undefined, true);
    let health;
    beforeAll(() => {
        health = new api_1.HealthAPI(avalanche);
    });
    afterEach(() => {
        jest_mock_axios_1.default.reset();
    });
    test("health", () => __awaiter(void 0, void 0, void 0, function* () {
        const result = health.health();
        const payload = {
            result: {
                checks: {
                    C: {
                        message: [Object],
                        timestamp: "2021-09-29T15:31:20.274427-07:00",
                        duration: 275539,
                        contiguousFailures: 0,
                        timeOfFirstFailure: null
                    },
                    P: {
                        message: [Object],
                        timestamp: "2021-09-29T15:31:20.274508-07:00",
                        duration: 14576,
                        contiguousFailures: 0,
                        timeOfFirstFailure: null
                    },
                    X: {
                        message: [Object],
                        timestamp: "2021-09-29T15:31:20.274529-07:00",
                        duration: 4563,
                        contiguousFailures: 0,
                        timeOfFirstFailure: null
                    },
                    isBootstrapped: {
                        timestamp: "2021-09-29T15:31:19.448314-07:00",
                        duration: 392,
                        contiguousFailures: 0,
                        timeOfFirstFailure: null
                    },
                    network: {
                        message: [Object],
                        timestamp: "2021-09-29T15:31:19.448311-07:00",
                        duration: 4866,
                        contiguousFailures: 0,
                        timeOfFirstFailure: null
                    },
                    router: {
                        message: [Object],
                        timestamp: "2021-09-29T15:31:19.448452-07:00",
                        duration: 3932,
                        contiguousFailures: 0,
                        timeOfFirstFailure: null
                    }
                },
                healthy: true
            }
        };
        const responseObj = {
            data: payload
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe(payload.result);
    }));
});
