// avalanche/typings/e2e_tests/e2etestlib.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTests = exports.Matcher = exports.getAvalanche = void 0;
const src_1 = require("src");
const getAvalanche = () => {
    if (typeof process.env.AVALANCHEGO_IP === "undefined") {
        throw "Undefined environment variable: AVALANCHEGO_IP";
    }
    if (typeof process.env.AVALANCHEGO_PORT === "undefined") {
        throw "Undefined environment variable: AVALANCHEGO_PORT";
    }
    const avalanche = new src_1.Avalanche(process.env.AVALANCHEGO_IP, parseInt(process.env.AVALANCHEGO_PORT));
    return avalanche;
};
exports.getAvalanche = getAvalanche;
var Matcher;
(function (Matcher) {
    Matcher[Matcher["toBe"] = 0] = "toBe";
    Matcher[Matcher["toEqual"] = 1] = "toEqual";
    Matcher[Matcher["toContain"] = 2] = "toContain";
    Matcher[Matcher["toMatch"] = 3] = "toMatch";
    Matcher[Matcher["toThrow"] = 4] = "toThrow";
    Matcher[Matcher["Get"] = 5] = "Get";
})(Matcher = exports.Matcher || (exports.Matcher = {}));
const createTests = (tests_spec) => {
    for (const [testName, promise, preprocess, matcher, expected] of tests_spec) {
        test(testName, () => __awaiter(void 0, void 0, void 0, function* () {
            if (matcher == Matcher.toBe) {
                expect(preprocess(yield promise())).toBe(expected());
            }
            if (matcher == Matcher.toEqual) {
                expect(preprocess(yield promise())).toEqual(expected());
            }
            if (matcher == Matcher.toContain) {
                expect(preprocess(yield promise())).toEqual(expect.arrayContaining(expected()));
            }
            if (matcher == Matcher.toMatch) {
                expect(preprocess(yield promise())).toMatch(expected());
            }
            if (matcher == Matcher.toThrow) {
                yield expect(preprocess(promise())).rejects.toThrow(expected());
            }
            if (matcher == Matcher.Get) {
                expected().value = preprocess(yield promise());
                expect(true).toBe(true);
            }
        }));
    }
};
exports.createTests = createTests;
