// avalanche/typings/tests/apis/info/api.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jest_mock_axios_1 = __importDefault(require("jest-mock-axios"));
const src_1 = require("src");
const bn_js_1 = __importDefault(require("bn.js"));
describe("Info", () => {
    const ip = "127.0.0.1";
    const port = 9650;
    const protocol = "https";
    const avalanche = new src_1.Avalanche(ip, port, protocol, 12345, "What is my purpose? You pass butter. Oh my god.", undefined, undefined, false);
    let info;
    beforeAll(() => {
        info = avalanche.Info();
    });
    afterEach(() => {
        jest_mock_axios_1.default.reset();
    });
    test("getBlockchainID", () => __awaiter(void 0, void 0, void 0, function* () {
        const result = info.getBlockchainID("X");
        const payload = {
            result: {
                blockchainID: avalanche.XChain().getBlockchainID()
            }
        };
        const responseObj = {
            data: payload
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe("What is my purpose? You pass butter. Oh my god.");
    }));
    test("getNetworkID", () => __awaiter(void 0, void 0, void 0, function* () {
        const result = info.getNetworkID();
        const payload = {
            result: {
                networkID: 12345
            }
        };
        const responseObj = {
            data: payload
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe(12345);
    }));
    test("getTxFee", () => __awaiter(void 0, void 0, void 0, function* () {
        const result = info.getTxFee();
        const payload = {
            result: {
                txFee: "1000000",
                creationTxFee: "10000000"
            }
        };
        const responseObj = {
            data: payload
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response.txFee.eq(new bn_js_1.default("1000000"))).toBe(true);
        expect(response.creationTxFee.eq(new bn_js_1.default("10000000"))).toBe(true);
    }));
    test("getNetworkName", () => __awaiter(void 0, void 0, void 0, function* () {
        const result = info.getNetworkName();
        const payload = {
            result: {
                networkName: "denali"
            }
        };
        const responseObj = {
            data: payload
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe("denali");
    }));
    test("getNodeID", () => __awaiter(void 0, void 0, void 0, function* () {
        const result = info.getNodeID();
        const payload = {
            result: {
                nodeID: "abcd"
            }
        };
        const responseObj = {
            data: payload
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe("abcd");
    }));
    test("getNodeVersion", () => __awaiter(void 0, void 0, void 0, function* () {
        const result = info.getNodeVersion();
        const payload = {
            result: {
                version: "avalanche/0.5.5"
            }
        };
        const responseObj = {
            data: payload
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe("avalanche/0.5.5");
    }));
    test("isBootstrapped false", () => __awaiter(void 0, void 0, void 0, function* () {
        const result = info.isBootstrapped("X");
        const payload = {
            result: {
                isBootstrapped: false
            }
        };
        const responseObj = {
            data: payload
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe(false);
    }));
    test("isBootstrapped true", () => __awaiter(void 0, void 0, void 0, function* () {
        const result = info.isBootstrapped("P");
        const payload = {
            result: {
                isBootstrapped: true
            }
        };
        const responseObj = {
            data: payload
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe(true);
    }));
    test("peers", () => __awaiter(void 0, void 0, void 0, function* () {
        const peers = [
            {
                ip: "127.0.0.1:60300",
                publicIP: "127.0.0.1:9659",
                nodeID: "NodeID-P7oB2McjBGgW2NXXWVYjV8JEDFoW9xDE5",
                version: "avalanche/1.3.2",
                lastSent: "2021-04-14T08:15:06-07:00",
                lastReceived: "2021-04-14T08:15:06-07:00",
                benched: null
            },
            {
                ip: "127.0.0.1:60302",
                publicIP: "127.0.0.1:9655",
                nodeID: "NodeID-NFBbbJ4qCmNaCzeW7sxErhvWqvEQMnYcN",
                version: "avalanche/1.3.2",
                lastSent: "2021-04-14T08:15:06-07:00",
                lastReceived: "2021-04-14T08:15:06-07:00",
                benched: null
            }
        ];
        const result = info.peers();
        const payload = {
            result: {
                peers
            }
        };
        const responseObj = {
            data: payload
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe(peers);
    }));
    test("uptime", () => __awaiter(void 0, void 0, void 0, function* () {
        const result = info.uptime();
        const uptimeResponse = {
            rewardingStakePercentage: "100.0000",
            weightedAveragePercentage: "99.2000"
        };
        const payload = {
            result: uptimeResponse
        };
        const responseObj = {
            data: payload
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toEqual(uptimeResponse);
    }));
});
